forge/src/debugger.rs

```diff
@@ -0,0 +1,268 @@
+                new_node.idx = self.arena.len();
+                new_node.location = self.arena[entry].children.len();
+                self.arena[entry].children.push(new_node.idx);
+                self.arena.push(new_node);
+            }
+            // We haven't found the parent node, go deeper
+            _ => self.push_node(
+                *self.arena[entry].children.last().expect("Disconnected debug node"),
+                new_node,
+            ),
+        }
+    }
+
+    /// Recursively traverses the tree of debug nodes and flattens it into a [Vec] where each
+    /// item contains:
+    ///
+    /// - The address of the contract being executed
+    /// - A [Vec] of debug steps along that contract's execution path
+    /// - A boolean denoting
+    /// Recursively traverses the tree of debug step nodes and flattens it into a
+    /// vector where each element contains
+    /// 1. the address of the contract being executed
+    /// 2. a vector of all the debug steps along that contract's execution path.
+    /// 3. Whether the contract was created in this node or not
+    ///
+    /// This makes it easy to pretty print the execution steps.
+    pub fn flatten(&self, entry: usize, flattened: &mut Vec<(Address, Vec<DebugStep>, bool)>) {
+        let node = &self.arena[entry];
+        flattened.push((node.address, node.steps.clone(), node.creation));
+        node.children.iter().for_each(|child| {
+            self.flatten(*child, flattened);
+        });
+    }
+}
+
+/// A node in the arena
+#[derive(Default, Debug, Clone)]
+pub struct DebugNode {
+    /// Parent node index in the arena
+    pub parent: Option<usize>,
+    /// Children node indexes in the arena
+    pub children: Vec<usize>,
+    /// Location in parent
+    pub location: usize,
+    /// This node's index in the arena
+    pub idx: usize,
+    /// Address context
+    pub address: Address,
+    /// Depth
+    pub depth: usize,
+    /// The debug steps
+    pub steps: Vec<DebugStep>,
+    /// Whether the contract was created in this node or not
+    pub creation: bool,
+}
+
+impl DebugNode {
+    pub fn new(address: Address, depth: usize, steps: Vec<DebugStep>) -> Self {
+        Self { address, depth, steps, ..Default::default() }
+    }
+}
+
+/// A `DebugStep` is a snapshot of the EVM's runtime state.
+///
+/// It holds the current program counter (where in the program you are),
+/// the stack and memory (prior to the opcodes execution), any bytes to be
+/// pushed onto the stack, and the instruction counter for use with sourcemap.
+#[derive(Debug, Clone)]
+pub struct DebugStep {
+    /// The program counter
+    pub pc: usize,
+    /// Stack *prior* to running the associated opcode
+    pub stack: Vec<H256>,
+    /// Memory *prior* to running the associated opcode
+    pub memory: Memory,
+    /// Opcode to be executed
+    pub instruction: Instruction,
+    /// Optional bytes that are being pushed onto the stack
+    pub push_bytes: Option<Vec<u8>>,
```

Comment:
I should adjust this to `Option<Bytes>` (same for call traces) since it's cheaper to clone